### PR TITLE
fix(scaleColumns): adjust for overflow when scaling columns in flex mode

### PR DESCRIPTION
**Pull request by peter-stoyanov**
**What kind of change does this PR introduce?** (check one with "x")

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

For some inputs the function scaleColumns(colsByGroup: any, maxWidth: any, totalFlexGrow: any) in utils/math.ts returns a set of column with total width exceeding that of the maxWidth argument.

This seems to be the reason for a horizontal scrollbar on large screens where space is more than enough.

**What is the new behavior?**

After auto-caling the columns in flex mode, a correction is made based on the outstanding overflow/underflow by adding/removing the delta from the biggest column.

**Does this PR introduce a breaking change?** (check one with "x")

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: projects/ngx-datatable/src/lib/utils/math.ts [hunk 1]
```diff
@@ -75,6 +75,27 @@ function scaleColumns(colsByGroup: any, maxWidth: any, totalFlexGrow: any) {
...
+    ref: null
+  };
+
+  for (const attr in colsByGroup) {
+    if (colsByGroup.hasOwnProperty(attr)) {
+      for (const column of colsByGroup[attr]) {
+        totalWidthAchieved += column.width;
+
+        if (column.width > biggestColumnRef.width) {
+          biggestColumnRef.ref = column
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/utils/math.ts:92]: Nit: missing `;`

```suggestion
          biggestColumnRef.ref = column;
```

File: projects/ngx-datatable/src/lib/utils/math.ts [hunk 2]
```diff
@@ -75,6 +75,27 @@ function scaleColumns(colsByGroup: any, maxWidth: any, totalFlexGrow: any) {
...
+        }
+      }
+    }
+  }
+  const remainingDelta = maxWidth - totalWidthAchieved;
+  biggestColumnRef.ref.width += remainingDelta;
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/utils/math.ts:98]: ` biggestColumnRef.ref` could potentially be `null`, so I guess we want to put an `if` around:

```suggestion
  if (biggestColumnRef.ref) {
    biggestColumnRef.ref.width += maxWidth - totalWidthAchieved;
  }
```

File: projects/ngx-datatable/src/lib/utils/math.spec.ts [hunk 3]
```diff
@@ -34,4 +34,27 @@ describe('Math function', () => {
...
+            { prop: 'id4', width: 175, maxWidth: undefined, minWidth: 175, flexGrow: 1, canAutoResize: true },
+            { prop: 'id5', width: 143, maxWidth: undefined, minWidth: 120, flexGrow: 1, canAutoResize: true }
+          ]
+        };
+
+        const givenTableWidth = 1180;
+
+        adjustColumnWidths(cols.center, givenTableWidth);
+
+        const totalAdjustedColumnWidths = cols.center.map(c => c.width).reduce((p, c) => p + c, 0)
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/utils/math.spec.ts:55]: Nit: missing `;`

```suggestion
        const totalAdjustedColumnWidths = cols.center.map(c => c.width).reduce((p, c) => p + c, 0);
```

File: projects/ngx-datatable/src/lib/utils/math.spec.ts [hunk 4]
```diff
@@ -34,4 +34,27 @@ describe('Math function', () => {
       });
     });
   });
+
+  describe('adjustColumnWidths', () => {
+    describe('flex mode', () => {
+      it('should not go overflow/underflow compared to given max width', () => {
+        const cols = {
+          center: [
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/utils/math.spec.ts:42]: Just out of curiosity: why do you additionally wrap the array here, when you always just access `cols.center`?

Otherwise, you could just go with a simple array here, as in all other tests:

```ts
const columns = [
  ...
```

**Review by fh1ch — CHANGES_REQUESTED** [review]
@peter-stoyanov thanks a ton for this PR, we appreciate this a lot :bow:

Really nice work, just some minor comments/findings from my end.

@timowolf @chintankavathia would you mind also giving this one a quick look? Thanks! It seems to be related to: https://github.com/swimlane/ngx-datatable/issues/2101

File: projects/ngx-datatable/src/lib/utils/math.spec.ts (hunk 4 above)
Comment (peter-stoyanov) [reply projects/ngx-datatable/src/lib/utils/math.spec.ts]: I didn't realise using groups is not required, I've simplified the test.

**Comment by peter-stoyanov** [comment]
Yes, it's related to https://github.com/swimlane/ngx-datatable/issues/2101, I didn't get any response there for more than a year and I just found out about this fork, decided to directly open a PR here.

File: projects/ngx-datatable/src/lib/utils/math.spec.ts (hunk 3 above)
Comment (peter-stoyanov) [reply projects/ngx-datatable/src/lib/utils/math.spec.ts]: Fixed

File: projects/ngx-datatable/src/lib/utils/math.ts (hunk 2 above)
Comment (peter-stoyanov) [reply projects/ngx-datatable/src/lib/utils/math.ts]: Fixed

File: projects/ngx-datatable/src/lib/utils/math.ts (hunk 1 above)
Comment (peter-stoyanov) [reply projects/ngx-datatable/src/lib/utils/math.ts]: Fixed

File: projects/ngx-datatable/src/lib/utils/math.ts [hunk 5]
```diff
@@ -75,6 +75,29 @@ function scaleColumns(colsByGroup: any, maxWidth: any, totalFlexGrow: any) {
...
+    width: 0,
+    ref: null
+  };
+
+  for (const attr in colsByGroup) {
+    if (colsByGroup.hasOwnProperty(attr)) {
+      for (const column of colsByGroup[attr]) {
+        totalWidthAchieved += column.width;
+
+        if (column.width > biggestColumnRef.width) {
```
Comment (spike-rabbit) [inline projects/ngx-datatable/src/lib/utils/math.ts:91]:
```suggestion
        if (column.canAutoResize && column.width > biggestColumnRef.width) {
```

**Review by spike-rabbit — CHANGES_REQUESTED** [review]
Thx a lot for contributing! @peter-stoyanov 

In addition to my comment please also check, that you are not violating any minWidths

File: projects/ngx-datatable/src/lib/utils/math.ts (hunk 5 above)
Comment (spike-rabbit) [reply projects/ngx-datatable/src/lib/utils/math.ts:91]: Unless I missed something, the previous implementation exceeded maxWidth in case `sum(minWidth) > maxWidth`.
I think we should keep this behavior.

Comment (peter-stoyanov) [reply projects/ngx-datatable/src/lib/utils/math.ts]: Ok, I've added checks for that, also refactored the logic a bit, given that if we violate minWidth on the biggest column, we may apply the correction on one of the other columns. I've also added two more unit tests, both with edge cases.

**Review by spike-rabbit — APPROVED** [review]
Looks good to me :+1:

**Review by fh1ch — APPROVED** [review]
@peter-stoyanov again, also from my end a huge thanks 🙇 

No new findings, so let's get this in.

LGTM 👍
